nl_server/embeddings.py

```diff
@@ -80,52 +131,201 @@ def get_embedding_at_index(self, index: int) -> List[float]:
+        for dcid in self.dcids[ent['corpus_id']].split(','):
+          # Prefer the top score.
+          if dcid not in query2sv2score[q]:
+            query2sv2score[q][dcid] = score
+            query2sv2sentence2score[q][dcid] = {}
+
+          if ent['corpus_id'] >= len(self.sentences):
+            continue
+          sentence = self.sentences[ent['corpus_id']]
+          query2sv2sentence2score[q][dcid][sentence] = score
+
+    query2result: Dict[str, VarCandidates] = {}
+
+    # Go over the map and prepare parallel lists of
+    # SVs and scores in query2result.
+    for q, sv2score in query2sv2score.items():
+      sv2score_sorted = [(k, v) for (
+          k,
+          v) in sorted(sv2score.items(), key=lambda item: item[1], reverse=True)
+                        ]
+      svs = [k for (k, _) in sv2score_sorted]
+      scores = [v for (_, v) in sv2score_sorted]
+      query2result[q] = VarCandidates(svs=svs, scores=scores, sv2sentences={})
+
+    # Go over the results and prepare the sv2sentences map in
+    # query2result.
+    for q, sv2sentence2score in query2sv2sentence2score.items():
+      query2result[q].sv2sentences = {}
+      for sv, sentence2score in sv2sentence2score.items():
+        query2result[q].sv2sentences[sv] = []
+        for sentence, score in sorted(sentence2score.items(),
+                                      key=lambda item: item[1],
+                                      reverse=True):
+          score = round(score, 4)
+          query2result[q].sv2sentences[sv].append(sentence + f' ({score})')
+
+    return query2result
+
+  #
+  # The main entry point to detect SVs.
+  #
+  def detect_svs(self, orig_query: str) -> Dict[str, Union[Dict, List]]:
+    # Remove all stop-words.
+    query_monovar = utils.remove_stop_words(orig_query,
+                                            query_util.ALL_STOP_WORDS)
+
+    # Search embeddings for a single SV.
+    result_monovar = self._search_embeddings([query_monovar])[query_monovar]
 
+    # Try to detect multiple SVs.  Use the original query so that
+    # the logic can rely on stop-words like `vs`, `and`, etc as hints
+    # for SV delimiters.
+    result_multivar = self._detect_multiple_svs(orig_query)
+
+    # TODO: Rename SV_to_Sentences for consistency.
     return {
-        'SV': svs_sorted,
-        'CosineScore': scores_sorted,
-        'EmbeddingIndex': sv_index_sorted,
-        'SV_to_Sentences': all_svs_sentences,
+        'SV': result_monovar.svs,
+        'CosineScore': result_monovar.scores,
+        'SV_to_Sentences': result_monovar.sv2sentences,
+        'MultiSV': _multivar_candidates_to_dict(result_multivar)
+    }
+
+  #
+  # Detects one or more SVs from the query.
+  # TODO: Fix the query upstream to ensure the punctuations aren't stripped.
+  #
+  def _detect_multiple_svs(self, query: str) -> MultiVarCandidates:
+    #
+    # Prepare a combination of query-sets.
+    #
+    querysets = query_util.prepare_multivar_querysets(query)
```

Comment:
Reading the comments below about a queryset, wouldn't it be better to use this combinatorial API/function built in to python? https://www.geeksforgeeks.org/itertools-combinations-module-python-print-possible-combinations/

Reply:
good idea.  doing in the follow on PR.